src/brod_consumer.erl

```diff
@@ -163,124 +142,88 @@ handle_info(init_socket, #state{ cb_mod     = CbMod
-  case brod_kafka:is_error(ErrorCode) of
-    true ->
-      Now = os:timestamp(),
-      Error = {ErrorCode, brod_kafka_errors:desc(ErrorCode)},
-      F = fun(CbStateIn) ->
-            CbMod:handle_error(Topic, Partition, Error, CbStateIn)
-          end,
-      State = State0#state{last_error = ?KAFKA_ERROR(Now, ErrorCode, F)},
-      maybe_handle_last_error(State);
-    false ->
-      case Messages of
-        [] ->
-          maybe_delay_fetch_request(State0#state.sleep_timeout),
-          {noreply, State0};
-        [_|_] ->
-          F = fun(CbStateIn) ->
-                  CbMod:handle_messages(Topic, Partition, HighWmOffset,
-                                        Messages, CbStateIn)
-              end,
-          CbPending = State0#state.cb_pending ++ [F],
-          CbPendingCnt = State0#state.cb_pending_cnt + 1,
-          State1 = State0#state{ offset = LastOffset + 1
-                               , cb_pending = CbPending
-                               , cb_pending_cnt = CbPendingCnt},
-          State2 = maybe_spawn_handle_messages(State1),
-          {ok, State} = maybe_send_fetch_request(State2),
-          {noreply, State}
-        end
+  State = State0#state{ socket_pid     = SocketPid
+                      , begin_offset   = Offset
+                      , max_wait_time  = MaxWaitTime
+                      , min_bytes      = MinBytes
+                      , max_bytes      = MaxBytes
+                      , sleep_timeout  = SleepTimeout
+                      , prefetch_count = PrefetchCount
+                      },
+  case resolve_begin_offset(State) of
+    {ok, NewState} ->
+      ok = brod_client:register_consumer(ClientPid, Topic, Partition),
+      {noreply, NewState};
+    {error, Reason} ->
+      {stop, {error, Reason}, State}
   end;
-%% handle obsolete fetch responses in case we got new offset from callback
-handle_info({msg, Pid, CorrId1, _R},
-            #state{last_corr_id = CorrId2, socket_pid = Pid} = State)
-  when CorrId1 < CorrId2 ->
-  error_logger:info_msg("~p ~p Dropping obsolete fetch response "
-                       "with corr_id = ~p",
-                       [?MODULE, self(), CorrId1]),
+handle_info({msg, _Pid, CorrId, R}, State) ->
+  handle_fetch_response(R, CorrId, State);
+handle_info(?SEND_FETCH_REQUEST, State0) ->
+  State = maybe_send_fetch_request(State0),
   {noreply, State};
-handle_info(?SEND_FETCH_REQUEST, State) ->
-  {ok, CorrId} = send_fetch_request(State),
-  {noreply, State#state{last_corr_id = CorrId}};
+handle_info({'DOWN', _MonitorRef, process, Pid, _Reason},
+            #state{subscriber = Pid} = State) ->
+  error_logger:info_msg("~p ~p subscriber ~p is down",
+                        [?MODULE, self(), Pid]),
+  NewState = reset_buffer(State#state{subscriber = undefined}),
+  {noreply, NewState};
 handle_info({'DOWN', _MonitorRef, process, Pid, Reason},
             #state{socket_pid = Pid} = State) ->
   {stop, {socket_down, Reason}, State};
```

Comment:
Shouldn't stop here, need to keep retrying.

I'll submit a PR with producer refactoring to show what I mean soon.